### ISSUE TITLE
NUnit3 Test Adapter not running all tests in VisualStudio 2015

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit3-vs-adapter#202]
@aspengler commented on [Wed Jul 06 2016](https://github.com/nunit/docs/issues/81)

Since the latest update of NUnit 3 Test Adapter (3.4.0), when pressing "Run All", only parts of the tests are executed (e.g. only the tests of the first project when grouped "By Project").

---

@CharliePoole commented on [Wed Jul 06 2016](https://github.com/nunit/docs/issues/81#issuecomment-230761667)

This is the repository for the NUnit documentation. Issues about the software should be filed in the repository for that software, in this case the NUnit3 Test adapter repo. I'm moving this there.

**Comment by CharliePoole** [comment]
What projects are involved - that is, what runtimes do they target? Are all the tests discovered? What shows up in the Output window?

**Comment by aspengler** [comment]
All tests (targeted at x64) are discovered, but only those tests in the first project in the list are executed.

The Output Window shows:

------ Run test started ------
NUnit Adapter 3.4.0.0: Test execution started
Running all tests in C:\Development\xxx\xxx\xxx\xxx\xxx\xxx\xxx\aaa.bbb.ccc\bin\Debug\aaa.bbb.ccc.Tests.dll
NUnit3TestExecutor converted 36 of 36 NUnit test cases
System.AppDomainUnloadedException: Attempted to access an unloaded AppDomain. This can happen if the test(s) started a thread but did not stop it. Make sure that all the threads started by the test(s) are stopped before completion.
Exception NUnit.Engine.NUnitEngineException, Exception thrown executing tests
Exception encountered unloading AppDomain
NUnit Adapter 3.4.0.0: Test execution complete
========== Run test finished: 36 run (0:02:44,413) ==========

**Comment by rprouse** [comment]
@CharliePoole I have started seeing the same error every test run since the 3.4 update. I only have one test project in my current solution, so maybe it is causing the subsequent assemblies to not run?

```
System.AppDomainUnloadedException: Attempted to access an unloaded AppDomain. This can happen if the test(s) started a thread but did not stop it. Make sure that all the threads started by the test(s) are stopped before completion.
Exception NUnit.Engine.NUnitEngineException, Exception thrown executing tests
Exception encountered unloading AppDomain
```

**Comment by CharliePoole** [comment]
This seems potentially critical. 

@rprouse What happens after you see the message? Can you run the tests? 

@aspengler I don't actually know what VS does when you have one X86 test and the rest are 64-bit. For it to work, VS would have to set up one process for the x86 and one or more for the x64. The adapter currently runs in the process it's given rather than examining the assembly and creating a process for it, so the assumption is that VS takes care of this. We need to run a few tests.

**Comment by aspengler** [comment]
The test projects are all x64...

**Comment by CharliePoole** [comment]
@aspengler OK... I misunderstood your comment.

It's possible we catch the exception in the wrong place - outside the loop on source assembly rather than inside it, for example. I'll check the code and try to reproduce this.

**Comment by aspengler** [comment]
Is there anything I could do? Debugging, ...?

**Comment by CharliePoole** [comment]
@aspengler Please give this build a try... If it resolves this it will become 3.4.1.

https://ci.appveyor.com/project/CharliePoole/nunit3-vs-adapter/build/3.5.0-ci-00236-issue-202/artifacts

**Comment by aspengler** [comment]
My colleague tried it out and it worked, appearantly.

Thanks, Charlie...
